unitTests/ui/AnalyticsTest.ts

```diff
@@ -158,6 +159,90 @@ export function AnalyticsTest() {
         });
       });
 
+      describe('with data layer in page', () => {
+        let test: Mock.IBasicComponentSetup<Analytics>;
+        let defaultDataLayerName = 'dataLayer';
+        let customDataLayerName = 'myDataLayer';
+        let wronglyConfiguredDataLayerName = 'badDataLayer';
+        let nonExistingDataLayerName = 'myImaginaryDataLayer';
+        let data = {
+          event: 'CoveoCustomEvent',
+          coveoAnalyticsEventData: {
+            language: 'en',
+            device: 'test device',
+            searchInterface: 'test interface',
+            searchHub: 'test searchHub',
+            responseTime: 100,
+            actionType: 'testActionType',
+            actionCause: 'testActionCause',
+            customMetadatas: { testKey: 'testValue' }
+          }
+        };
+        beforeEach(() => {
+          (<any>window)[defaultDataLayerName] = [];
+          (<any>window)[customDataLayerName] = [];
+          (<any>window)[wronglyConfiguredDataLayerName] = [];
+          (<any>window)[wronglyConfiguredDataLayerName].push = 'April Fools!';
+        });
+
+        afterEach(() => {
+          delete (<any>window)[wronglyConfiguredDataLayerName];
+          delete (<any>window)[customDataLayerName];
+          delete (<any>window)[defaultDataLayerName];
+        });
+
+        it('should not push to default data layer if pushToGtmDataLayer is false', () => {
+          test = Mock.basicComponentSetup<Analytics>(Analytics);
+          $$(test.env.root).trigger(AnalyticsEvents.analyticsEventReady, data);
+          expect((<any>window)[defaultDataLayerName].length).toBe(0);
+        });
+
+        it('should not push to default data layer if pushToGtmDataLayer is true and gtmDataLayerName is the empty string', () => {
+          test = Mock.optionsComponentSetup<Analytics, IAnalyticsOptions>(Analytics, {
+            pushToGtmDataLayer: true,
+            gtmDataLayerName: ''
+          });
+          $$(test.env.root).trigger(AnalyticsEvents.analyticsEventReady, data);
+          expect((<any>window)[defaultDataLayerName].length).toBe(0);
+        });
+
+        it('should not push to undefined data layer if pushToGtmDataLayer is true', () => {
+          test = Mock.optionsComponentSetup<Analytics, IAnalyticsOptions>(Analytics, {
+            pushToGtmDataLayer: true,
+            gtmDataLayerName: nonExistingDataLayerName
+          });
+          $$(test.env.root).trigger(AnalyticsEvents.analyticsEventReady, data);
+          expect(typeof (<any>window)[nonExistingDataLayerName]).toBe('undefined');
+        });
+
+        it('should catch error attempting to push to invalid data layer', () => {
```

Comment:
This one doesn't work (in fact, it always succeeds, even when it shouldn't).

I'm leaving it in the PR so you guys can help me out :)

Reply:
When you say it does not work, I see the test is checking that calling `trigger` does not throw an error. Isn't that what we would expect considering the condition on line 497 in Analytics.ts?

```
if (!(<any>window)[dataLayerName]) {
    this.logger.warn(`Cannot push to Google Tag Manager data layer: '${dataLayerName}' is undefined.`);
    return false;
}
```

Would it be better to expect `expect((<any>window)[defaultDataLayerName][0]).toBe(undefined);` instead?

Reply:
In this test, I do define the data layer, but overwrite its `push` property so it is no longer a function.

I would expect triggering the AnalyticsEventReady event not to throw in such a case, because the `(<any>window)[dataLayerName].push` call is wrapped in a try-catch.

It does not throw right now... but even if I remove the try-catch and rebuild, the test still passes, so it's apparently invalid.

Reply:
I see, I would make sure that
1) the code is calling `handleAnalyticsEventReady` in Analytics.ts
2) That there is no code that is overwriting/resetting what you have stored on  window[dataLayerName].
3) That the window in your test is the same window in the source file.